clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -549,7 +571,8 @@ void pollSelectionKeys(Set<SelectionKey> selectionKeys,
                 }
 
                 /* if channel is ready write to any sockets that have space in their buffer and for which we have data */
-                if (channel.ready() && key.isWritable()) {
+                if (channel.ready() && key.isWritable()
+                        && !channel.maybeBeginClientReauthentication(time.nanoseconds())) {
```

Comment:
I don't think we can use `currentTimeNanos` because it represents the time when `poll()` was invoked, and that could be as far back in time as the timeout value (i.e. a large timeout value could in theory result in `java.nio.channels.Selector.select()` blocking for quite a while).  If we use a time to far in the past it increases the chance of the reauth decision yielding `false` when in fact it should yield `true` -- and in that case we could end up with our connection being killed.

What we can do is reuse the `channelStartTimeNanos` if it is available (i.e. `channelStartTimeNanos != 0 ? channelStartTimeNanos : time.nanoseconds()`) since it is a very recent value.  Furthermore, we can delay actually getting the time as long as possible -- and therefore maybe we don't need to calculate it at all -- by making `KafkaChannel` accept a `Supplier<Long>` rather than a `long`.  I've made both of these changes.

Reply:
Actually, I just realized that `currentTimeNanos` does not represent the time when `poll()` was invoked -- it actually represents the moment after the `select()` returns.  So we can use that value if necessary.